drivers/firmware/scmi/core.c

```diff
@@ -176,6 +171,62 @@ int scmi_send_message(struct scmi_protocol *proto, struct scmi_message *msg,
 	}
 }
 
+int scmi_read_message(struct scmi_protocol *proto, struct scmi_message *msg)
+{
+	int ret = 0;
+
+	if (!proto->rx) {
+		return -ENODEV;
+	}
+
+	if (!proto->rx->ready) {
+		return -EINVAL;
+	}
+
+	/* read message from platform
+	 * such as notification event and give response for it
+	 */
+	ret = scmi_transport_read_message(proto->transport, proto->rx, msg);
+	if (ret < 0) {
+		return ret;
+	}
+
+	if (k_is_pre_kernel()) {
+		/* send p2a response into platform */
+		ret = scmi_transport_send_message(proto->transport, proto->rx, msg);
+		if (ret < 0) {
+			LOG_ERR("failed to send message");
+			return ret;
+		}
+	} else {
+		/* wait for channel to be free */
+		ret = k_mutex_lock(&proto->rx->lock, K_USEC(SCMI_CHAN_LOCK_TIMEOUT_USEC));
+		if (ret < 0) {
+			LOG_ERR("failed to acquire chan lock");
+			return ret;
```

Comment:
Here you are returning with mutex locked. I think you might want to `goto out_release_mutex`.